### PR TITLE
Use HashSet to eliminiate duplicate declared folder and files

### PR DESCRIPTION
**Pull request by troydai**
To address the DNX issue https://github.com/aspnet/dnx/issues/1588

/cc @lodejard @loudej @davidfowl

### CODE REVIEW

**Comment by troydai** [comment]
I'm running a few more performance test to verify the fix.

File: src/Microsoft.Framework.FileSystemGlobbing/Internal/MatcherContext.cs [hunk 1]
```diff
@@ -17,8 +16,8 @@ public class MatcherContext
         private readonly IList<IPatternContext> _excludePatternContexts;
         private readonly IList<string> _files;
 
-        private readonly IList<LiteralPathSegment> _declaredLiteralFolderSegments = new List<LiteralPathSegment>();
-        private readonly IList<LiteralPathSegment> _declaredLiteralFileSegments = new List<LiteralPathSegment>();
+        private readonly ISet<LiteralPathSegment> _declaredLiteralFolderSegments = new HashSet<LiteralPathSegment>();
+        private readonly ISet<LiteralPathSegment> _declaredLiteralFileSegments = new HashSet<LiteralPathSegment>();
```
Comment (davidfowl) [inline src/Microsoft.Framework.FileSystemGlobbing/Internal/MatcherContext.cs]: Why not just make both sides HashSet if they are assigned in place?

Comment (troydai) [inline src/Microsoft.Framework.FileSystemGlobbing/Internal/MatcherContext.cs]: All right.

File: src/Microsoft.Framework.FileSystemGlobbing/Internal/PathSegments/LiteralPathSegment.cs [hunk 2]
```diff
@@ -18,5 +18,18 @@ public bool Match(string value, StringComparison comparisonType)
         {
             return string.Equals(Value, value, comparisonType);
         }
+
+        public override bool Equals(object obj)
+        {
+            var other = obj as LiteralPathSegment;
+
+            return other != null &&
+                other.Value.Equals(Value, StringComparison.Ordinal);
```
Comment (Eilon) [inline src/Microsoft.Framework.FileSystemGlobbing/Internal/PathSegments/LiteralPathSegment.cs]: Use static `string.Equals`?

**Comment by troydai** [comment]
/cc @Praburaj 

@davidfowl @Eilon Updated :watch:

**Comment by davidfowl** [comment]
Why is the CI failing?

**Comment by Praburaj** [comment]
Change looks good to me. Do you also need some tests for case sensitivity. Eg. "sub/source.cs" and "Sub/source.cs" etc.

**Comment by troydai** [comment]
@Praburaj case-sensitive case will come in when I fix the directory case sensitive case.

**Comment by troydai** [comment]
@davidfowl CI failed at package restore.

File: test/Microsoft.Framework.FileSystemGlobbing.Tests/FunctionalTests.cs [hunk 3]
```diff
@@ -29,6 +29,17 @@ public void Dispose()
         }
 
         [Fact]
+        public void DuplicatePatterns()
```
Comment (davidfowl) [inline test/Microsoft.Framework.FileSystemGlobbing.Tests/FunctionalTests.cs]: Shouldn't this test both types of slashes?

Comment (troydai) [inline test/Microsoft.Framework.FileSystemGlobbing.Tests/FunctionalTests.cs]: There are tests added to test slash and backslash in the input patterns. Those tests are added when tests on *nix was fixed.

Comment (davidfowl) [inline test/Microsoft.Framework.FileSystemGlobbing.Tests/FunctionalTests.cs]: Use the attributes that @victorhurdugaci wrote. Have windows specific tests

Comment (troydai) [inline test/Microsoft.Framework.FileSystemGlobbing.Tests/FunctionalTests.cs]: Why? all the tests should pass on both Windows and *nix. I can add variation for backslash in this test. But a conditional test for Windows only shouldn't exist. We suppose to take both types of slashes.

Comment (davidfowl) [inline test/Microsoft.Framework.FileSystemGlobbing.Tests/FunctionalTests.cs]: Why would paths with `\` work on nix isn't that kinda weird?

Comment (troydai) [inline test/Microsoft.Framework.FileSystemGlobbing.Tests/FunctionalTests.cs]: Absolute path is not allowed. There are tests coverage ensuring the exceptions are thrown.

The path `\` can be very possible when a project is developed on Windows and move to *nix unless we abolish all the usage of `\`.

**Comment by troydai** [comment]
:watch:

File: src/Microsoft.Framework.FileSystemGlobbing/Internal/PathSegments/LiteralPathSegment.cs [hunk 4]
```diff
@@ -18,5 +18,18 @@ public bool Match(string value, StringComparison comparisonType)
...
+        {
+            var other = obj as LiteralPathSegment;
+
+            return other != null &&
+                string.Equals(other.Value, Value, StringComparison.Ordinal);
+        }
+
+        public override int GetHashCode()
+        {
+            return Value.GetHashCode();
```
Comment (victorhurdugaci) [inline src/Microsoft.Framework.FileSystemGlobbing/Internal/PathSegments/LiteralPathSegment.cs]: The value can be null...

Comment (troydai) [inline src/Microsoft.Framework.FileSystemGlobbing/Internal/PathSegments/LiteralPathSegment.cs]: :+1 I'll add guard to Value to ensure it's not null.

Comment (troydai) [inline src/Microsoft.Framework.FileSystemGlobbing/Internal/PathSegments/LiteralPathSegment.cs]: Will call into `base.GetHashCode()`

**Comment by troydai** [comment]
CI passed. Any further comments, everyone?

:watch:

**Comment by victorhurdugaci** [comment]
:shipit:

**Pull request merged**